src/linalg.jl

```diff
@@ -0,0 +1,92 @@
+import Base: transpose!, permutedims!
+# function transpose_kernel!(
+#         state, At, A::AbstractArray{T}, width, height, ::Val{BLOCK}, ::Val{LMem}
+#     ) where {BLOCK, LMem, T}
+#
+#     ui1 = UInt32(1)
+#     bidx_x = blockidx_x(state) - ui1
+#     bidx_y = blockidx_y(state) - ui1
+#     tidx_x = threadidx_x(state) - ui1
+#     tidx_y = threadidx_y(state) - ui1
+#
+#     A_local = @LocalMemory(state, T, LMem)
+#
+#     base_idx_a = bidx_x * BLOCK + bidx_y * (BLOCK * width)
+#     base_idx_a_t = bidx_y * BLOCK + bidx_x * (BLOCK * height)
+#
+#     glob_idx_a = base_idx_a + tidx_x + width * tidx_y
+#     glob_idx_a_t = base_idx_a_t + tidx_x + height * tidx_y
+#     glob_idx_a >= length(A) && return
+#     A_local[tidx_y * BLOCK + tidx_x + ui1] = A[glob_idx_a + ui1]
+#     synchronize_threads(state)
+#     At[glob_idx_a_t + ui1] = A_local[tidx_x * BLOCK + tidx_y + ui1]
+#     return
+# end
+
+function transpose_blocks!(
+        state, odata::AbstractArray{T}, idata, ::Val{SHMEM}, ::Val{TDIM}, ::Val{BLOCK_ROWS}, ::Val{NROW}
+    ) where {T, SHMEM, TDIM, BLOCK_ROWS, NROW}
+
+    ui1 = Cuint(1)
```

Comment:
Can we not call this `Cuint`? :P

Reply:
sure... I've been mixing Cuint/UInt32, which isn't particularly nice.